torch/distributed/rpc/api.py

```diff
@@ -202,6 +203,74 @@ def _all_gather(obj):
     return states.gathered_objects
 
 
+# detect if any worker has invalid map_location configurations, and return
+# names of failed workers
+@_require_initialized
+def _setup_map_locations(map_locations):
+    agent = _get_current_rpc_agent()
+    if not map_locations:
+        return
+    elif not isinstance(agent, TensorPipeAgent):
+        raise ValueError(
+            "map_locations attribute only works with TensorPipe RPC Backend, "
+            f"but the current agent type is {type(agent)}"
+        )
+
+    def check_one_worker(name, map_locations, all_device_counts):
+        device_count = all_device_counts[name]
+        wrong_worker_names = set(map_locations) - set(all_device_counts)
+        if wrong_worker_names:
+            raise ValueError(f"Wrong worker names: {wrong_worker_names}")
+        for worker_name in all_device_counts:
+            remote_device_count = all_device_counts[worker_name]
+            if worker_name in map_locations:
+                map_location = map_locations[worker_name]
+                key_set = set(map_location.keys())
+                val_set = set(map_location.values())
+                if not all([
+                    len(map_location) == len(key_set),
+                    len(map_location) == len(val_set),  # check 1-to-1 mapping
+                    min(key_set) >= -1,
+                    max(key_set) < device_count,  # check local range
+                    min(val_set) >= -1,
+                    max(val_set) < remote_device_count  # check remote range
+                ]):
+                    raise ValueError(
+                        f"Invalid map_location configuration on {name}:\n"
+                        f"map_locations = {map_locations}"
+                    )
+
+                if -1 not in map_location and -1 in map_location.values():
+                    raise ValueError(
+                        f"Invalid map_location configuration on {name}. "
+                        "It maps a non-CPU device to CPU, when combined with "
+                        "the default CPU-CPU mapping, device mapping for CPU "
+                        "tensors is ambiguous:\n"
+                        f"map_locations = {map_locations}"
+                    )
+
+
+    gathered = _all_gather([torch.cuda.device_count(), map_locations])
+    all_device_counts = {name: gathered[name][0] for name in gathered}
+    all_map_locations = {name: gathered[name][1] for name in gathered}
+    for worker_name in all_map_locations:
+        worker_map_locations = all_map_locations[worker_name]
+        check_one_worker(worker_name, worker_map_locations, all_device_counts)
+
+    # passed all checked, construct reverse mapping for return values
+    reverse_map_locations = {}
```

Comment:
> I still think that the all-gather is an overkill for this though, because it means that worker #K will also receive the maps between workers #N and #M, even though it doesn't need them. So there is extra data being sent around. What we would need, in principle, is a pairwise swap, so basically an all-to-all collective. But I don't think it makes sense to implement it, let's just roll with all-gather for now.

Let me add an `all_to_all` in the follow up PR. @gqchen would like to know how RPC `all_to_all` compares to c10d `all_to_all` as well. So adding it here would serves two requirements, which seems worth it.

Reply:
TBH, implementing all-to-all just because of this seems like an overkill: we can stick to the quadratic solution until it becomes a problem and then re-evaluate. If we end up implementing an all-to-all because of other reasons then, at that point, we could consider it here, but we shouldn't make that block the progress on this PR.